examples/with-metadata.rs

```diff
@@ -0,0 +1,146 @@
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0
+            .metadata()
+            .inner
+            .get()
+            .partial_cmp(&other.0.metadata().inner.get())
+            .map(|ord| ord.reverse())
+    }
+}
+
+impl Ord for ByDuration {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.0
+            .metadata()
+            .inner
+            .get()
+            .cmp(&other.0.metadata().inner.get())
+            .reverse()
+    }
+}
+
+pin_project! {
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    struct MeasureRuntime<'a, F> {
+        #[pin]
+        f: F,
+        duration: &'a Cell<Duration>
+    }
+}
+
+impl<'a, F: Future> Future for MeasureRuntime<'a, F> {
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let duration_cell = this.duration;
+        let start = Instant::now();
+        let res = F::poll(this.f, cx);
+        let new_duration = Instant::now() - start;
+        duration_cell.set(duration_cell.get() / 2 + new_duration / 2);
+        res
+    }
+}
+
+pub struct DurationMetadata {
+    inner: Cell<Duration>,
+}
+
+thread_local! {
+    // A queue that holds scheduled tasks.
+    static QUEUE: RefCell<BinaryHeap<ByDuration>> = RefCell::new(BinaryHeap::new());
+}
+
+fn make_future_fn<'a, F>(
+    future: F,
+) -> impl (FnOnce(&'a DurationMetadata) -> MeasureRuntime<'a, F>) {
+    move |duration_meta| MeasureRuntime {
+        f: future,
+        duration: &duration_meta.inner,
+    }
+}
+
+fn ensure_safe_schedule<F: Send + Sync + 'static>(f: F) -> F {
+    f
+}
+
+/// Spawns a future on the executor.
+pub fn spawn<F, T>(future: F) -> Task<T, DurationMetadata>
+where
+    F: Future<Output = T> + 'static,
+    T: 'static,
+{
+    let spawn_thread_id = thread::current().id();
+    // Create a task that is scheduled by pushing it into the queue.
+    let schedule = ensure_safe_schedule(move |runnable| {
+        if thread::current().id() != spawn_thread_id {
+            panic!("Task would be run on a different thread than spawned on.");
+        }
+        QUEUE.with(move |queue| queue.borrow_mut().push(ByDuration(runnable)));
+    });
+    let future_fn = make_future_fn(future);
+    let (runnable, task) = unsafe {
+        Builder::new()
+            .metadata(DurationMetadata {
+                inner: Cell::new(Duration::default()),
+            })
+            .spawn_unchecked(future_fn, schedule)
+    };
+
+    // Schedule the task by pushing it into the queue.
+    runnable.schedule();
+
+    task
+}
+
+pub fn block_on<F>(future: F)
+where
+    F: Future<Output = ()> + 'static,
+{
+    let task = spawn(future);
+    while !task.is_finished() {
+        let Some(runnable) = QUEUE.with(|queue| queue.borrow_mut().pop()) else { thread::yield_now(); continue };
+        runnable.0.run();
+    }
+}
```

Comment:
Why not use `smol::block_on()`?

Reply:
Because all the futures submitted are `!Send`, so the executor has to be driven within the thread.

Reply:
`smol::block_on()` does not send futures to other threads, it polls it exclusively in the current thread.

Reply:
Right, but I'm saying that using `smol::block_on` would mean there's nothing popping off of `QUEUE` and running the runnables, so it would just hang.

Reply:
Maybe instead change it to an `async run()` method, and use `smol`'s channels to carry the runnables so you can `.await` on a new runnable?

Reply:
I think this is roughly the same as the `run` method within the `spawn-local` example as-is, except instead of using a flume and detaching we just directly check `is_finished` for the spawned task.